### PR TITLE
AI AttackOrFleeFuzzy improvements

### PR DESCRIPTION
**Pull request by reaperrr**
I needed a break from the other stuff I'm working on, so I gave this a look.

The idea behind the string shortenings is the assumption that the FuzzyLibrary must be doing *some* kind of string comparison, so shorter strings should speed things up a little (likely insignificant, but who knows...).

The other commits should, at least in theory, make the AI behave a little smarter when making attack-or-flee decisions.

### CODE REVIEW

**Review by chrisforbes — DISMISSED** [review]

**Comment by reaperrr** [comment]
Minor update to 2nd commit, pulled the .Clamp out of the foreach and added a comment detailing the reasons for the clamp values.

**Review by GraionDilach — APPROVED** [review]
:+1:

File: OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs [hunk 1]
```diff
@@ -221,8 +221,8 @@ static float RelativePower(IEnumerable<Actor> own, IEnumerable<Actor> enemy)
 				var arms = a.TraitsImplementing<Armament>();
 				foreach (var arm in arms)
 				{
-					var warhead = arm.Weapon.Warheads.OfType<DamageWarhead>().FirstOrDefault();
-					if (warhead != null)
+					var dmgWarheads = arm.Weapon.Warheads.OfType<DamageWarhead>();
```
Comment (pchote) [inline OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs]: `damageWarheads` please.

File: OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs [hunk 2]
```diff
@@ -25,7 +25,7 @@ sealed class AttackOrFleeFuzzy
 		{
 			"if ((OwnHealth is Normal) " +
 			"and ((EnemyHealth is NearDead) or (EnemyHealth is Injured) or (EnemyHealth is Normal)) " +
-			"and ((RelativeAttackPower is Weak) or (RelativeAttackPower is Equal) or (RelativeAttackPower is Strong)) " +
+			"and ((RelAtkPwr is Weak) or (RelAtkPwr is Equal) or (RelAtkPwr is Strong)) " +
```
Comment (pchote) [inline OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs]: I don't buy that these would make any measurable performance difference, and IMO the shortened versions are a lot less readable.  Can we please drop these renames?

File: OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs [hunk 3]
```diff
@@ -23,104 +23,92 @@ sealed class AttackOrFleeFuzzy
 	{
 		static readonly string[] DefaultRulesNormalOwnHealth = new[]
 		{
-			"if ((OwnHealth is Normal) " +
-			"and ((EnemyHealth is NearDead) or (EnemyHealth is Injured) or (EnemyHealth is Normal)) " +
-			"and ((RelativeAttackPower is Weak) or (RelativeAttackPower is Equal) or (RelativeAttackPower is Strong)) " +
-			"and ((RelativeSpeed is Slow) or (RelativeSpeed is Equal) or (RelativeSpeed is Fast))) " +
+			"if ((OwnHP is Normal) " +
+			"and ((EnemyHP is NearDead) or (EnemyHP is Injured) or (EnemyHP is Normal)) " +
+			"and ((RelAtkPwr is Weak) or (RelAtkPwr is Equal) or (RelAtkPwr is Strong)) " +
```
Comment (pchote) [inline OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs]: I don't buy that these would make any measurable performance difference, and IMO the shortened versions are a lot less readable.  Can we please drop these renames?

File: OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs [hunk 4]
```diff
@@ -221,9 +209,16 @@ static float RelativePower(IEnumerable<Actor> own, IEnumerable<Actor> enemy)
...
-					if (warhead != null)
-						sumOfDamage += warhead.Damage;
+					var burst = arm.Weapon.Burst;
+
+					// For simplicity's sake, we're only factoring in the first burst delay, as more than one burst delay is extremely rare.
+					// Additionally, clamping total delay to minimum of 1 (ReloadDelay: 0 is technically possible) and maximum of 200.
+					// High dmg/low ROF weapons shouldn't be rated too low as high dmg/shot can outweigh mere dps due to likelier 1-hit-kills.
+					// TODO: Revisit this at some point to replace the arbitrary cap with something smarter.
+					var totalReloadDelay = arm.Weapon.ReloadDelay + (arm.Weapon.BurstDelays[0] * (burst - 1)).Clamp(1, 200);
+					var dmgWarheads = arm.Weapon.Warheads.OfType<DamageWarhead>();
```
Comment (pchote) [inline OpenRA.Mods.Common/AI/AttackOrFleeFuzzy.cs]: `damageWarheads` please.

**Comment by reaperrr** [comment]
Updated.

**Review by pchote — APPROVED** [review]
Code changes look sensible and I didn't notice any obvious regressions ingame.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/fec14d143977fcd44623c003cf6cbf21210a2baf

**Pull request merged**